server/src/main/java/org/apache/iotdb/db/engine/upgrade/UpgradeTask.java

```diff
@@ -40,30 +47,69 @@ public UpgradeTask(TsFileResource upgradeResource) {
   @Override
   public void runMayThrow() {
     try {
-      upgradeResource.getWriteQueryLock().readLock().lock();
-      String tsfilePathBefore = upgradeResource.getFile().getAbsolutePath();
-      String tsfilePathAfter = UpgradeUtils.getUpgradeFileName(upgradeResource.getFile());
-
-      UpgradeLog.writeUpgradeLogFile(
-          tsfilePathBefore + COMMA_SEPERATOR + UpgradeCheckStatus.BEGIN_UPGRADE_FILE);
+      List<TsFileResource> upgradedResources = generateUpgradedFiles();
       upgradeResource.getWriteQueryLock().writeLock().lock();
+      String oldTsfilePath = upgradeResource.getFile().getAbsolutePath();
       try {
-        FSFactoryProducer.getFSFactory().getFile(tsfilePathBefore).delete();
-        FSFactoryProducer.getFSFactory()
-            .moveFile(FSFactoryProducer.getFSFactory().getFile(tsfilePathAfter),
-                FSFactoryProducer.getFSFactory().getFile(tsfilePathBefore));
+        // delete old TsFile
+        upgradeResource.remove();
+        // move upgraded TsFiles to their own partition directories
+        for (TsFileResource upgradedResource : upgradedResources) {
+          File upgradedFile = upgradedResource.getFile();
+          long partition = upgradedResource.getTimePartitionWithCheck();
+          String storageGroupPath = upgradedFile.getParentFile().getParentFile().getParent();
+          File partitionDir = FSFactoryProducer.getFSFactory().getFile(storageGroupPath, partition + "");
+          if (!partitionDir.exists()) {
+            partitionDir.mkdir();
+          }
+          FSFactoryProducer.getFSFactory().moveFile(upgradedFile,
+              FSFactoryProducer.getFSFactory().getFile(partitionDir, upgradedFile.getName()));
+          upgradedResource.setFile(
+              FSFactoryProducer.getFSFactory().getFile(partitionDir, upgradedFile.getName()));
+          upgradedResource.serialize();
+          // delete tmp partition folder when it is empty
+          if (upgradedFile.getParentFile().isDirectory() 
+              && upgradedFile.getParentFile().listFiles().length == 0) {
```

Comment:
this may produce NullPointerException, better to have a check

Reply:
Isn't checking `isDirectory()` enough? 
The javadoc of listFiles said 
`If this abstract pathname does not denote a directory, then this
method returns null .  Otherwise an array of File objects
is returned, one for each file or directory in the directory.`